src/memray/reporters/tui.py

```diff
@@ -1,83 +1,165 @@
+import contextlib
+import dataclasses
 import os
+import pathlib
+import sys
+import threading
 from collections import defaultdict
 from collections import deque
-from copy import deepcopy
 from dataclasses import dataclass
 from datetime import datetime
+from functools import total_ordering
 from math import ceil
+from typing import Any
 from typing import DefaultDict
 from typing import Deque
 from typing import Dict
 from typing import Iterable
 from typing import List
 from typing import Optional
-from typing import Sequence
 from typing import Set
 from typing import Tuple
+from typing import cast
 
-from rich.layout import Layout
 from rich.markup import escape
-from rich.panel import Panel
-from rich.progress_bar import ProgressBar
-from rich.table import Column
-from rich.table import Table
+from rich.segment import Segment
+from rich.style import Style
+from rich.text import Text
+from textual import events
+from textual import log
+from textual.app import App
+from textual.app import ComposeResult
+from textual.binding import Binding
+from textual.color import Color
+from textual.color import Gradient
+from textual.containers import Container
+from textual.containers import HorizontalScroll
+from textual.dom import DOMNode
+from textual.message import Message
+from textual.reactive import reactive
+from textual.screen import Screen
+from textual.strip import Strip
+from textual.widget import Widget
+from textual.widgets import DataTable
+from textual.widgets import Footer
+from textual.widgets import Label
+from textual.widgets import Static
+from textual.widgets.data_table import RowKey
 
 from memray import AllocationRecord
+from memray import SocketReader
 from memray._memray import size_fmt
 
 MAX_MEMORY_RATIO = 0.95
 
 DEFAULT_TERMINAL_LINES = 24
 
 
-class MemoryGraph:
+@dataclass(frozen=True)
+class Location:
+    function: str
+    file: str
+
+
+@dataclass
+class AllocationEntry:
+    own_memory: int
+    total_memory: int
+    n_allocations: int
+    thread_ids: Set[int]
+
+
+@dataclass(frozen=True, eq=False)
+class Snapshot:
+    heap_size: int
+    records: List[AllocationRecord]
+    records_by_location: Dict[Location, AllocationEntry]
+
+
+_EMPTY_SNAPSHOT = Snapshot(heap_size=0, records=[], records_by_location={})
+
+
+class SnapshotFetched(Message):
+    def __init__(self, snapshot: Snapshot, disconnected: bool) -> None:
+        self.snapshot = snapshot
+        self.disconnected = disconnected
+        super().__init__()
+
+
+class MemoryGraph(Widget):
+    lines: reactive[Tuple[str, ...]] = reactive(())
```

Comment:
Is this used for something?

Reply:
See line 526 of `tui.py`.

Reply:
Yeah I can see that the Graph is used there but I cannot see where the "lines" attribute is used

Reply:
Nope, that must be a holdover from a previous iteration. The tests pass and the TUI works even with it removed.

Reply:
🤦 I misread the comment location, sorry.

Reply:
It wouldn't be surprising if there's more of these. This went through a few rewrites between the initial and final version, so it's possible some dead code got missed here and there.